### PR TITLE
rest: fix some id search issue

### LINKED ISSUE #450: rest: operator like with uuid won't work
### How to reproduce your problem

do a resource search with `id like "foobar"`

### What is the result that you get

DEBUG [pifpaf.drivers] uwsgi[11888] output:   File "/workspace/sileht/os_dev/stack/gnocchi/client/.tox/py27/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1398, in _handle_dbapi_exception
DEBUG [pifpaf.drivers] uwsgi[11888] output:     util.raise_from_cause(newraise, exc_info)
DEBUG [pifpaf.drivers] uwsgi[11888] output:   File "/workspace/sileht/os_dev/stack/gnocchi/client/.tox/py27/local/lib/python2.7/site-packages/sqlalchemy/util/compat.py", line 203, in raise_from_cause
DEBUG [pifpaf.drivers] uwsgi[11888] output:     reraise(type(exception), exception, tb=exc_tb, cause=cause)
DEBUG [pifpaf.drivers] uwsgi[11888] output:   File "/workspace/sileht/os_dev/stack/gnocchi/client/.tox/py27/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1182, in _execute_context
DEBUG [pifpaf.drivers] uwsgi[11888] output:     context)
DEBUG [pifpaf.drivers] uwsgi[11888] output:   File "/workspace/sileht/os_dev/stack/gnocchi/client/.tox/py27/local/lib/python2.7/site-packages/sqlalchemy/engine/default.py", line 470, in do_execute
DEBUG [pifpaf.drivers] uwsgi[11888] output:     cursor.execute(statement, parameters)
DEBUG [pifpaf.drivers] uwsgi[11888] output: DBError: (psycopg2.ProgrammingError) operator does not exist: uuid ~~ uuid
DEBUG [pifpaf.drivers] uwsgi[11888] output: LINE 3: WHERE resource.id LIKE 'b709a41a-fe8e-5070-83d8-7ae45d62b5bb...

### What is result that you expected

An error 400 and not an error 500

### LINKED ISSUE #491: rest: broken search query id+like+list
```
     - name: search like id
       POST: /v1/search/resource/generic
       data:
         like:
           id:
             - fa%
       status: 400
```

      File "/workspace/sileht/os_dev/stack/gnocchi/gnocchi/.tox/py27-postgresql-file/local/lib/python2.7/site-packages/voluptuous/validators.py", line 262, in __call__
        v = schema(v)
      File "/workspace/sileht/os_dev/stack/gnocchi/gnocchi/.tox/py27-postgresql-file/local/lib/python2.7/site-packages/voluptuous/schema_builder.py", line 221, in __call__
        return self._compiled([], data)
      File "/workspace/sileht/os_dev/stack/gnocchi/gnocchi/.tox/py27-postgresql-file/local/lib/python2.7/site-packages/voluptuous/schema_builder.py", line 538, in validate_dict
        return base_validate(path, iteritems(data), out)
      File "/workspace/sileht/os_dev/stack/gnocchi/gnocchi/.tox/py27-postgresql-file/local/lib/python2.7/site-packages/voluptuous/schema_builder.py", line 315, in validate_mapping
        cval = cvalue(key_path, value)
      File "/workspace/sileht/os_dev/stack/gnocchi/gnocchi/.tox/py27-postgresql-file/local/lib/python2.7/site-packages/voluptuous/schema_builder.py", line 716, in validate_callable
        return schema(data)
      File "gnocchi/rest/api.py", line 1049, in ResourceUUID
        return utils.ResourceUUID(value, creator)
      File "gnocchi/utils.py", line 50, in ResourceUUID
        return uuid.UUID(value)
      File "/usr/lib/python2.7/uuid.py", line 133, in __init__
        hex = hex.replace('urn:', '').replace('uuid:', '')
    AttributeError: 'list' object has no attribute 'replace'

### PR DESCRIPTION
**Pull request by sileht**
id key and like operator have some issue, due to the binary nature of
uuid.

This change avoids it.

Closes #450
Closes #491

### CODE REVIEW

File: gnocchi/rest/api.py [hunk 1]
```diff
@@ -1293,11 +1300,16 @@ def _ResourceSearchSchema():
...
                     u">=", u"≥", u"ge",
                     u"!=", u"≠", u"ne",
-                    u"like"
                 ): voluptuous.All(
                     voluptuous.Length(min=1, max=1),
                     {"id": _ResourceUUID,
-                     six.text_type: ResourceSearchSchemaAttributeValue},
+                     NotIDKey: ResourceSearchSchemaAttributeValue},
+                ),
+                voluptuous.Any(
```
Comment (jd) [inline gnocchi/rest/api.py]: It does not look like you need `voluptuous.Any` here if it's just to pass one value.

File: gnocchi/tests/functional/gabbits/search.yaml [hunk 2]
```diff
@@ -40,6 +40,21 @@ tests:
        - "expected a list for dictionary value @ data["
        - "'and']"
 
+    - name: search like id
+      POST: /v1/search/resource/generic
+      data:
+        like:
+          id: fa%
+      status: 400
```
Comment (jd) [inline gnocchi/tests/functional/gabbits/search.yaml]: Can you check the details of the error here? So we're sure it's the right one caught and it's explicit.

File: gnocchi/tests/functional/gabbits/search.yaml [hunk 3]
```diff
@@ -40,6 +40,21 @@ tests:
...
+
+    - name: search like list id
+      POST: /v1/search/resource/generic
+      data:
+        like:
+          id:
+            - fa%
+      status: 400
```
Comment (jd) [inline gnocchi/tests/functional/gabbits/search.yaml]: ditto

**Comment by sileht** [comment]
Pull-request updated, HEAD is now b609b3f48bccb1a5cb30ab015987519750c92f5c

File: gnocchi/rest/api.py [hunk 4]
```diff
@@ -1302,20 +1311,21 @@ def _ResourceSearchSchema():
                     u"<=", u"≤", u"le",
                     u">=", u"≥", u"ge",
                     u"!=", u"≠", u"ne",
-                    u"like"
                 ): voluptuous.All(
                     voluptuous.Length(min=1, max=1),
                     {"id": _ResourceUUID,
-                     six.text_type: ResourceSearchSchemaAttributeValue},
+                     NotIDKey: ResourceSearchSchemaAttributeValue},
```
Comment (jd) [inline gnocchi/rest/api.py]:
```python
>>> voluptuous.NotIn(["id"])("foo")
'foo'
>>> voluptuous.NotIn(["id"])("id")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/site-packages/voluptuous/validators.py", line 676, in __call__
    raise NotInInvalid(self.msg or 'value is not allowed')
voluptuous.error.NotInInvalid: value is not allowed
```
WDYT?

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 570f81cd6c8e555d38e78cb7ea37008ec602a279

File: gnocchi/rest/api.py (hunk 4 above)
Comment (sileht) [reply gnocchi/rest/api.py]: Well, I may need glasses I have looks at voluptuous and haven't found this....

**Comment by sileht** [comment]
I have used NotIn, it's there: https://github.com/gnocchixyz/gnocchi/pull/492/files#diff-26833c83685c0e5ed47a117f885c62a9R1289

**Pull request merged**